include/deal.II/grid/grid_tools.h

```diff
@@ -1285,6 +1389,272 @@ namespace GridTools
+        send_components;
+
+      /**
+       * Ranks to send to.
+       */
+      std::vector<unsigned int> send_ranks;
+
+      /**
+       * Information of each received data value. The elements of the tuple are
+       * as follows: 0) rank of sender, 1) local index, 2) found intersections.
+       *
+       * @note The vector is sorted according to 1), 0), 2).
+       *
+       * @note Multiple intersections between cells can be found
+       */
+      std::vector<std::tuple<unsigned int, unsigned int, IntersectionAnswer>>
+        recv_components;
+
+      /**
+       * Pointers of ranges to found intersections for requested intersection.
+       */
+      std::vector<unsigned int> recv_ptrs;
+
+      /**
+       * Ranks from where data is received.
+       */
+      std::vector<unsigned int> recv_ranks;
+
+      /**
+       * Distribute quadrature points to found intersections and construct
+       * GridTools::internal::DistributedComputePointLocationsInternal from
+       * class members. This can be done without searching for points agains
+       * since all information is locally known.
+       */
+      GridTools::internal::DistributedComputePointLocationsInternal<dim,
+                                                                    spacedim>
+      convert_to_distributed_compute_point_locations_internal(
+        const unsigned int                  n_quadrature_pnts,
+        const Triangulation<dim, spacedim> &tria,
+        const Mapping<dim, spacedim> &      mapping) const
+      {
+        using CellIterator =
+          typename Triangulation<dim, spacedim>::active_cell_iterator;
+
+        GridTools::internal::DistributedComputePointLocationsInternal<dim,
+                                                                      spacedim>
+          result;
+
+
+        // We need quadrature rules for the intersections. We are using a
+        // QGaussSimplex quadrature rule since CGAL always returns simplices as
+        // intersections.
+        const QGaussSimplex<dim - 1> quadrature(n_quadrature_pnts);
+
+        // Resulting quadrature points get different indices. In the case the
+        // requested intersections are unique also the resulting quadrature
+        // points are unique and we can simply number the points in an
+        // ascending way.
+        for (auto const &recv_component : recv_components)
+          {
+            // dependent on the size of the intersection an empty quadrature is
+            // returned. Therefore, we have to compute the quadrature also here.
+            const auto &quad = quadrature.compute_affine_transformation(
+              std::get<2>(recv_component));
+
+            for (unsigned int i = 0; i < quad.size(); ++i)
+              {
+                result.recv_components.emplace_back(
+                  std::get<0>(recv_component),
+                  result.recv_components.size(), // number of point
+                  numbers::invalid_unsigned_int);
```

Comment:
Why `numbers::invalid_unsigned_int`?

Reply:
```DistributedComputePointLocationsInternal``` stores an index related to the communication at the third position. This index is determined during the sorting and initialization of the CRS pointers, and therefore we don't have to care about it while initially filling recv_components.

Reply:
I added a comment in [3acdad4](https://github.com/dealii/dealii/pull/15165/commits/3acdad41005bde0d56374766d18c34a455fedc0a)